src/index.js

```diff
@@ -27,38 +30,90 @@ const warningObject = {
 };
 const ReduxStoreContext = createContext(warningObject);
 
-// helper hooks
+// utils
 
-const forcedReducer = state => state + 1;
-const useForceUpdate = () => useReducer(forcedReducer, 0)[1];
+const createMap = (keys, create) => {
+  // "Map" here means JavaScript Object not JavaScript Map.
+  const obj = {};
+  for (let i = 0; i < keys.length; ++i) {
+    const key = keys[i];
+    obj[key] = create(key);
+  }
+  return obj;
+};
 
-const useProxyfied = (state) => {
-  // cache
-  const proxyMap = useRef(new WeakMap());
-  const trappedMap = useRef(new WeakMap());
-  // trapped
+const createTrapped = (state, cache) => {
   let trapped;
-  if (trappedMap.current.has(state)) {
-    trapped = trappedMap.current.get(state);
+  if (cache && cache.trapped.has(state)) {
+    trapped = cache.trapped.get(state);
     trapped.reset();
   } else {
-    trapped = proxyState(state, null, proxyMap.current);
-    trappedMap.current.set(state, trapped);
+    trapped = proxyState(state, null, cache && cache.proxy);
+    if (cache) cache.trapped.set(state, trapped);
   }
-  // update ref
-  const lastProxyfied = useRef(null);
-  useLayoutEffect(() => {
-    lastProxyfied.current = {
-      state,
-      affected: trapped.affected,
-    };
+  return trapped;
+};
+
+const deproxifyResult = (object) => {
+  if (typeof object !== 'object') return object;
+  if (isProxyfied(object)) return deproxify(object);
+  const result = Array.isArray(object) ? [] : {};
+  let altered = false;
+  Object.key(object).forEach((key) => {
+    result[key] = deproxifyResult(object[key]);
+    if (object[key] !== result[key]) {
```

Comment:
In this case it does mean that you are returning a piece of the original object, and this `key` should go to `resultAffected`, and be used in `proxyCompare`(every key matter).
Key in mind - the `key` would be not on the "surface".

PS: dont use `key`, you need the original `suffix`, and you can get it back from proxy
```js
    const key = getProxyKey(object);
    resultAffected.push(key.suffix);
```
The original `addAffected` function from `addAffected` are a bit more complex, cos it tracking more that original object simultanetuly.

Reply:
https://github.com/theKashey/memoize-state/blob/master/src/call.js#L12

Reply:
I don't think I get `resultAffected`. What is it?
Or even, what is `deproxifyResult` in memoize-state? That can be something I misunderstand.

Reply:
```js
const selectTodos = state => ({
  todos: state.todos,
  amount: state.todos.length
});
```
Done - it's broken. A "deeper" access to `state.todos.length` cheated `getValuables` so `state.todos` is not _valuable_ anymore.

But then
```js
deproxify(selectTodos(state));
....
result[key] = deproxifyResult(object[key]);
if (object[key] !== result[key]) {
// object.todos !== result.todos /// 🧐, why it could be?
getProxyKey(object.todos); 
// aaahh! So you are `state.todos`, long time, no see 
```
Like you are getting object "as-is", thus could not use `collectValuables` to track it - it's _valuiable_ "as-is", as should be tracked "as-is".
In short - use `proxyEqual`/`proxyShallowEqual` to compare `normal affected`, and use `proxyCompare` to compare `resultAffected`.

Here is a better example, why it could be important
```js
const mapStateToProps = state => ({
   bill: getBill(state),
   totalAfterTax: getTotalAfterTax(state),
   currency: getCurrency(state);
});

where `getTotalAfterTax = getBill().total`, but __you dont know about it__, selectors are abstracting _structural access_ from you.

const Component = () => (
  <RenderBill bill={bill} /> <--- you are passing object as in deeper
  <Total total={totalAfterTax} currency={currency} /> 
)
```
^ thats broken, if bill got updated, but total is the same - it would not update.

Reply:
Oh my... I wasn't even close.
Thanks for the explanation.
Let me think about it...

Reply:
Just leaving a comment for now: I don't feel like doing that part in deep in my code unless it is really needed. So, I should probably use memoize-state (I've been thinking about it). Will learn more about it later.